### ISSUE TITLE
Improve search framework

### ISSUE THREAD

**Issue by jonasraoni** [pkp/pkp-lib#8920]

**Comment by diegoabadan** [comment]
Hello!

To our knowledge, an institution with the recently released OJS 3.4.0-7 and which hosts several journals was experiencing immense slowness and contacted us. We had to apply this fix:
https://forum.pkp.sfu.ca/t/ojs-3-4-0-4-performance-issue-ojs-3-3-0-14-to-ojs-3-4-0-4/83655/5?u=abadan

**Comment by asmecher** [comment]
@ewhanson -- just a heads-up that we're interested in OpenSearch as an option.

**Comment by asmecher** [comment]
I've implemented a first cut at replacing the built-in search index (and hook-based option for other implementations) using Laravel Scout as a middle layer. You can specify and configure driver in `config.inc.php`:
```
;;;;;;;;;;;;;;;;;;;
; Search Settings ;
;;;;;;;;;;;;;;;;;;;

[search]

; The search driver to use. Options: database (default), opensearch
driver = database

; For indexing engines supporting it (e.g. opensearch), provide a name for the submission index.
search_index_name = "submissions"

; When using the OpenSearch driver, the following must be specified...
;opensearch_hosts = "[\"https://localhost:9200\"]"
;opensearch_username = "admin"
;opensearch_password = "password"
;opensearch_ssl_verification = On
```

The `database` driver (implemented in `lib/pkp/classes/search/engines/DatabaseEngine.php`) uses fulltext indexing in MySQL, MariaDB, or PostgreSQL. This seems to be widely enough supported that we can make use of it now.

The `opensearch` driver (implemented in `lib/pkp/classes/search/engines/OpenSearchEngine.php`), once you configure it in `config.inc.php`, doesn't need any additional action on the OpenSearch side -- just run `tools/rebuildSearchIndex.php` to create and build the index.

(We can't use the built-in Scout engines because they presume we use Eloquent for submission content, and they don't match our use case very well. But the Scout engines are nice and simple so maintaining our own isn't nearly the overhead we had with our previous implementations.)

I would like to review and improve some things, but consider this a good enough "first cut" to get merged into `main` so later improvements can happen there iteratively.

@jonasraoni, would you mind taking a first review of the PRs linked in the issue description for general approach? Then I can port to OMP and OPS, and work through the set of issues referred to in the head of the list to make sure we have good coverage of the full set of requests/features.

(The "recommend by similar" and "recommend by author" plugins could both use further work, but I consider that out of scope here. See e.g. [#9897](https://github.com/pkp/pkp-lib/issues/9897).)

**Comment by asmecher** [comment]
Thanks for the reviews, @jonasraoni -- I've worked through and resolved just about all the comments and left notes on the remainder. I'll go ahead and port this over to OMP and OPS.

**Comment by asmecher** [comment]
@Tribunal33, over to you for testing. Functionally there should be no difference from the previous search toolset; this is a back-end rewrite.

**Comment by Tribunal33** [comment]
@asmecher I took a look on Jarda's VPS for main and it is no longer throwing a 500 error but I'm not seeing what the search is doing?  I did a basic search for a published article title and it returns no results (see image below).  I used the dropdowns for advanced search to find a particular date range of articles that are published and no results were displayed.  

<img width="1281" height="774" alt="Image" src="https://github.com/user-attachments/assets/9152ef15-d292-4d12-a104-50dcfaefde05" />

Some of the previous search expectations that I could find are a bit old and lack a lot of details [search docs from v2](https://docs.pkp.sfu.ca/learning-ojs-2/en/searching_submissions)

Here are some of the typical search requirements I'm looking for:

- [ ] Basic search functionality (what is searchable in an article?).  I'll assume that we would only want results that are published but maybe not.
- [ ] Are there any operators that can be used? AND or OR or symbols like && or ||?
- [ ] Are those advanced dropdowns used independently or as a filter for the main search?
- [ ] In a multi journal install would articles from other journals be available?  I would think not but could see benefits for a reader to expose other journals.  
- [ ] During an upgraded install what are the before states and after states during a migration?  Does our current dataset show this transition going from 3.5 to 3.6?
- [ ] How often does the index refresh and what is the triggers that refresh the index?
- [ ] In terms of the problems that this new framework solves I see some non-functional expectations.  How does it improve the accuracy, how much did it improve the performance, or how much storage is saved.

**Comment by asmecher** [comment]
@Tribunal33 I found an issue with that specific submission because the workflow creates then unpublishes a publication. The unpublish action was deleting all entries for the submission when it should've left the earlier, already-published version in place. [This commit](https://github.com/pkp/pkp-lib/commit/065a844088ccc953b47f0bb7051f9999d62fa97a) should fix it -- once the datasets rebuild in OJS you should be able to verify.

**Comment by Tribunal33** [comment]
From what I understood from our meeting @asmecher was that this ticket could be broken up into two parts.  Part 1.  Some general clean up of existing code and improve the search functionality.  And Part 2. Add new possibility for OpenSearch.

I think for passing this ticket we can do some basic functionality checks for both part 1 and part 2.  There is no testing list so understanding its previously existing functionality isn't possible.  Perhaps once we move to a phase 2 implementation we can work on some more robust testing list.  So far these are some functionalities/enhancements I have seen:

- [ ] By default all content from all Journals are displayed upon page load.  Unspecified ordering.
- [ ] When there is a multi Journal install the base index search will need to be added through admin > site navigation or you can append /search to the index url.
- [ ] In a multi Journal install a new field dropdown can focus the search to specific Journal or when left blank will search all journals.
- [ ] Advanced Search By Author has been removed (it didn't work as far as I could tell anyways)
- [ ] With Author being removed the URL search query parameter should be removed as well
- [ ] Advanced Search By Published After and Before can be used independently of the main search field.
- [ ] It appears words like AND or OR can be used with expected results.
- [ ] The main search field will only search for published articles
- [ ] The main search field will be able to search on the index of Title, Abstract, Author, and Keyword
- [ ] The index is re-compiled when a new article is published or you run the script to reindex in the /tools directory
- [ ] Keywords additionally can be searched when added to the URL by adding `keyword[]=science` to search for keyword science.
- [ ] Search main field and advanced filters should main their state when navigating through pagination.
- [ ] URL query parameters should be updated / maintained when navigating the pagination pages

OpenSearch requirements
- [ ] Similar to above search functionality
- [ ] By default all content from all Journals are displayed upon page load. Unspecified ordering
- [ ] System configuration setting is reflected under Seach > Driver > OpenSearch

**Comment by Tribunal33** [comment]
@asmecher I did some preliminary search tests on ORE with OpenSearch as the search driver and noticed a few issues.

1. Doesn't seem to be able to search by author but I could without OpenSearch.

<img width="859" height="467" alt="Image" src="https://github.com/user-attachments/assets/202e11ae-9c50-4bbb-95e1-f76a10ae75ac" />

2. Doesn't seem to search by keyword when input through the main search field.  I was able to use the URL `keywords[]=` and it did work
3. The results total doesn't seem to be correct.  I've noticed this with our default installs too.  I believe there are some articles being counted but not displayed.

<img width="906" height="260" alt="Image" src="https://github.com/user-attachments/assets/f1532b41-0ebf-442d-8780-fb9c6398b0a7" />

4. This one is more of an added question of what is the specified default list order of each of these search drivers?
a.  I noticed that on the /index page it lists the Latest Publications and I can assume by published date, but this list is different from the search page list where I'm not sure of the ordering.
b. When you reload the /search page the default list can sometimes flip around
sometimes this is the top most result

<img width="893" height="121" alt="Image" src="https://github.com/user-attachments/assets/980e952c-ec63-4ae8-a6a3-351e41732a42" />

sometimes this one

<img width="880" height="119" alt="Image" src="https://github.com/user-attachments/assets/168e4c87-100b-4dff-aac6-d5b00228d2e1" />

c.  Will this ordering change depending upon what driver is being used?  Does it even matter?

**Comment by Tribunal33** [comment]
@asmecher do you think we can close this one and maybe create a second part to improving search functionality and address the requirements I put in this comment : https://github.com/pkp/pkp-lib/issues/8920#issuecomment-3212386646 and the OpenSearch issues mentioned here : https://github.com/pkp/pkp-lib/issues/8920#issuecomment-3216001923 ?  We can include the #8435 issue a sub-issue.

**Comment by jardakotesovec** [comment]
@Tribunal33 @asmecher Here is adjustment to re-expose date variables to the template to fix UI&links on search page.   After code review I would check if the same thing needs to happen for omp/ops


**EDIT**
OJS: https://github.com/pkp/ojs/pull/5166
OMP: https://github.com/pkp/omp/pull/2162
OPS: https://github.com/pkp/ops/pull/1132

**Comment by Tribunal33** [comment]
Thanks @jardakotesovec.  Let me know when it is ready to test.  Once it passes QA I will remove it from the list of expectations as noted above.

**Comment by asmecher** [comment]
@jardakotesovec, thanks, I added a [brief review](https://github.com/pkp/ojs/pull/5166#pullrequestreview-3425107010).

**Comment by jardakotesovec** [comment]
@Tribunal33 Merged now.
Fixes:
1) OJS - dates UI now works same as 3.5, and are not lost in pagination links
2) OPS - same as OJS, but also removed the author section as thats not currently supported by the search (and was removed from OJS intentionally)
3) OMP - thats the most basic search, only small fix there to show search phrase in the input, when its in the query params. Which basically means not loosing it after clicking on search / pagination

**Comment by asmecher** [comment]
PRs to homogenize the category browse and search interfaces, in order to permit search parameters in the category browse:

- https://github.com/pkp/pkp-lib/pull/12199
- https://github.com/pkp/ojs/pull/5267
- https://github.com/pkp/omp/pull/2208
- https://github.com/pkp/ops/pull/1179
- https://github.com/asmecher/frascati/pull/2

Not ready for review or merge yet!

**Comment by asmecher** [comment]
@jardakotesovec, I've merged [these PRs](https://github.com/pkp/pkp-lib/issues/8920#issuecomment-3741328170) to `main`, which add support for search filtering in the "browse category" interface. For example, to browse a category while also searching for a keyword in the results metadata, you would use a URL like this:
```
http://.../index.php/publicknowledge/en/catalog/category/anthropology?keywords[]=keyword1
```

**Comment by asmecher** [comment]
@jardakotesovec and @ewhanson, I've put together a quick proposal to add search faceting API endpoints for keyword and subject metadata. This will be used in OSS ORE to support the advanced search filters. Could I have some early comments from each of you?

- Erik, this is a little different than our other API patterns (e.g. we are not nesting it below an entity), but I think it makes sense here. What do you think?
- Jarda, does this suit the needs for the theme? i.e. listing matches for keywords and subjects, with an optional keyword filter. (Does it need to return *all* values, or just the top matches?)

An example API:
```
http://.../index.php/ore/api/v1/search/facets/keywords?filter=health
```
...where you can use `keywords` or `subjects`, and `filter=health` is an optional keyword filter.

Example results:
```json
{"items":{"health":6,"mental health":6,"One Health":5,"Public Health":4}}
```
...where the numbers indicate the number of matching records (publications).

PRs:
- OJS: https://github.com/pkp/ojs/pull/5284
- pkp-lib: https://github.com/pkp/pkp-lib/pull/12235

**Comment by ewhanson** [comment]
Hey @asmecher, I think the API endpoint in this context makes sense, and I don't see any problems with having it be a search-specific one and not mapped to a particular entity. I imagine other search-related endpoints would probably need to follow a similar route.

I'm happy to have a look at some API style/implementation details when the PRs are out of draft as well.

One question I had about the overall response structure: how will the facets and the counts be used by other API endpoints, searching, UI, etc?

It may also be useful to return some data about the total number if items available vs the number of items returned in the cases where the available items exceed the maximum facet count of 100.

**Comment by ewhanson** [comment]
Also given our use of Laravel Scout and our current needs for faceted search, etc., do you see us having a more complete search API as part of OJS, or will it be limited to the backend PHP to actually perform and handle the searching?

**Comment by asmecher** [comment]
Thanks, @ewhanson!

> How will the facets and the counts be used by other API endpoints, searching, UI, etc?

They won't be used by other API endpoints; they'll be used to populate the advanced search fields, in order to allow users to filter search results (and browse results in the Categories interface) by keywords, subjects, and other facetable fields.

> Also given our use of Laravel Scout and our current needs for faceted search, etc., do you see us having a more complete search API as part of OJS, or will it be limited to the backend PHP to actually perform and handle the searching?

I think we probably will add a search REST API at some point, but we don't have a requirement for that now.

Laravel Scout gives us the benefit of an [interface between layers](https://laravel.com/docs/12.x/scout#custom-engines) that can support multiple back ends (OpenSearch and database fulltext at the moment), but it doesn't support much in the way of advanced queries -- boolean searches, nested logic, anything like that. My preference is that we focus on a single, high quality search field (and a few basic filters) that provides results in a good ordering with the best at the top; this aligns with general trends in search away from complex forms, and Scout will happily meet our needs. The database search engine will do OK, but richer toolsets like OpenSearch will prioritize match fields, take care of stemming, etc.

**Comment by asmecher** [comment]
After meeting with @jardakotesovec and @Godoy0722, I'm opting not to merge the [faceting endpoint](https://github.com/pkp/pkp-lib/issues/8920#issuecomment-3786747236) PRs for now -- we have adequate search field coverage for what we need (particularly once https://github.com/ajnyga/funding/pull/68 is merged, adding indexing in OpenSearch for funders), and the facet sets will be fetched server-side when the page is built. We don't currently need a separate API endpoint for this -- but the code in the PRs can be used to give examples of how these can be assembled server-side.

**Comment by asmecher** [comment]
@Tribunal33, could you summarize (or better yet, just link to) any unmet feedback from the QA work you did on this? All dev work is done from my perspective; some of the features won't be exposed for testing until the theming work is complete, but I'm hoping we can handle that in the form of smaller single issues and knock off this mega-issue.

**Comment by Tribunal33** [comment]
@asmecher I'm happy with moving this to completed as it was an enhancement task.  Hard to know what the expectations are for the improvement when I'm not sure how it behaved previously.  There are no prior tests to validate with.  I, however, posted a list of some typical expectations that went unanswered.

For example from my comments above:

- [ ] Basic search functionality (what is searchable in an article?).  I'll assume that we would only want results that are published but maybe not. <-- I think @ajnyga suggested unpublished keywords could be bad for example.  What about PMUR published states would that count?
- [ ] Are there any operators that can be used? AND or OR or symbols like && or ||?
- [ ] Are those advanced dropdowns used independently or as a refinement filter for the main search?
- [ ] In a multi journal install would articles from other journals be available for search?  I would think not but could see benefits for a reader to expose other journals.  On the other hand cross-journal polination could be bad for keywords.  
- [ ] During an upgraded install what are the before states and after states during a migration?  Does our current dataset show this transition going from 3.5 to 3.6?  Meaning if I search for something in 3.5 will it be assumed to be there in newer version or does the new version have to build the index again?
- [ ] How often does the index refresh and what are the triggers that refresh the index to add new searchable keywords?
- [ ] In terms of the problems that this new framework solves I see some non-functional expectations.  How does it improve the accuracy, how much did it improve the performance, or how much storage is saved.  Are these testable.

**Comment by jonasraoni** [comment]
As this issue is still open, I'll leave a reminder here.

Looks like deleting a submission doesn't clear its data from the OpenSearch driver.

**Comment by jonasraoni** [comment]
Another issue regarding length limits:

```php
OpenSearch\Common\Exceptions\BadRequest400Exception: {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Document contains at least one immense term in field=\"authors.en.keyword\" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[66, 101, 114, 110, 104, 97, 114, 100, 32, 72, 97, 117, 115, 100, 111, 114, 102, 32, 69, 108, 105, 99, 105, 111, 32, 84, 97, 112, 105, 97]...', original message: bytes can be at most 32766 in length; got 39509"}],"type":"illegal_argument_exception","reason":"Document contains at least one immense term in field=\"authors.en.keyword\" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[66, 101, 114, 110, 104, 97, 114, 100, 32, 72, 97, 117, 115, 100, 111, 114, 102, 32, 69, 108, 105, 99, 105, 111, 32, 84, 97, 112, 105, 97]...', original message: bytes can be at most 32766 in length; got 39509","caused_by":{"type":"max_bytes_length_exceeded_exception","reason":"bytes can be at most 32766 in length; got 39509"}},"status":400} in /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Connections/Connection.php:680
Stack trace:
#0 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Connections/Connection.php(338): OpenSearch\Connections\Connection->process4xxError()
#1 /var/www/html/lib/pkp/lib/vendor/react/promise/src/Internal/FulfilledPromise.php(47): OpenSearch\Connections\Connection->OpenSearch\Connections\{closure}()
#2 /var/www/html/lib/pkp/lib/vendor/ezimuel/ringphp/src/Future/CompletedFutureValue.php(66): React\Promise\Internal\FulfilledPromise->then()
#3 /var/www/html/lib/pkp/lib/vendor/ezimuel/ringphp/src/Core.php(341): GuzzleHttp\Ring\Future\CompletedFutureValue->then()
#4 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Connections/Connection.php(270): GuzzleHttp\Ring\Core::proxy()
#5 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Connections/Connection.php(248): OpenSearch\Connections\Connection->OpenSearch\Connections\{closure}()
#6 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Transport.php(117): OpenSearch\Connections\Connection->performRequest()
#7 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/LegacyTransportWrapper.php(34): OpenSearch\Transport->performRequest()
#8 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Client.php(2182): OpenSearch\LegacyTransportWrapper->sendRequest()
#9 /var/www/html/lib/pkp/lib/vendor/opensearch-project/opensearch-php/src/OpenSearch/Client.php(1858): OpenSearch\Client->performRequest()
#10 /var/www/html/lib/pkp/classes/search/engines/OpenSearchEngine.php(135): OpenSearch\Client->create()
#11 /var/www/html/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php(275): PKP\search\engines\OpenSearchEngine->PKP\search\engines\{closure}()
#12 /var/www/html/lib/pkp/classes/search/engines/OpenSearchEngine.php(69): Illuminate\Support\Collection->each()
#13 /var/www/html/lib/pkp/classes/observers/listeners/UpdateSubmissionInSearchIndex.php(51): PKP\search\engines\OpenSearchEngine->update()
#14 /var/www/html/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(512): PKP\observers\listeners\UpdateSubmissionInSearchIndex->handlePublicationPublished()
#15 /var/www/html/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(319): Illuminate\Events\Dispatcher->Illuminate\Events\{closure}()
#16 /var/www/html/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(299): Illuminate\Events\Dispatcher->invokeListeners()
#17 /var/www/html/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php(505): Illuminate\Events\Dispatcher->dispatch()
#18 /var/www/html/lib/pkp/classes/publication/Repository.php(706): event()
#19 /var/www/html/plugins/importexport/articleImporter/parsers/jats/PublicationParser.php(215): PKP\publication\Repository->publish()
#20 /var/www/html/plugins/importexport/articleImporter/BaseParser.php(127): APP\plugins\importexport\articleImporter\parsers\jats\Parser->getPublication()
#21 /var/www/html/plugins/importexport/articleImporter/ArticleVersion.php(143): APP\plugins\importexport\articleImporter\BaseParser->execute()
#22 /var/www/html/plugins/importexport/articleImporter/ArticleEntry.php(135): APP\plugins\importexport\articleImporter\ArticleVersion->process()
#23 /var/www/html/plugins/importexport/articleImporter/ArticleImporterPlugin.php(124): APP\plugins\importexport\articleImporter\ArticleEntry->process()
#24 /var/www/html/tools/importExport.php(78): APP\plugins\importexport\articleImporter\ArticleImporterPlugin->executeCLI()
#25 /var/www/html/tools/importExport.php(83): importExport->execute()
```

**Comment by asmecher** [comment]
@jonasraoni, hmm, do you have any information about which submission this was? A set of author names > 32k long seems pretty unusual!

**Comment by asmecher** [comment]
@jonasraoni, on [your comment](https://github.com/pkp/pkp-lib/issues/8920#issuecomment-4445521031):

> Looks like deleting a submission doesn't clear its data from the OpenSearch driver.

The submission should be deleted from the OpenSearch index on an "unpublish" event. How are you removing submissions? (Are they unpublished?)

**Comment by asmecher** [comment]
Closing as resolved --but still interested in feedback on the above two comments, @jonasraoni!

**Comment by jonasraoni** [comment]
I've checked the articles that failed to be indexed and we've got a huge list of authors (which will probably need some fixes), I've just wanted to ensure that the limit for the authors isn't a kind of "shared limit" (e.g. abstract limited to 32kb).

The submissions were deleted using the `Repo::submission()->delete()`, I didn't investigate anything though, I'll have to delete things again later, I'll add a reminder to double check if this is still happening :)
